Added new OpenTelemetry post (#227)
ippontech/blog-usa: posts/opentelemetry-and-friends.md

![x-ray-diagram](https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2022/11/opentelemetry_collector_diagram.png)

The OTel Collector, at a high level, can be divided into [three main parts](https://opentelemetry.io/docs/collector/configuration/):
- Receivers: Can, via endpoints, receive data from applications, for example using `otlpreceiver`, or they can be configured in a pull mode to scrape data from an application, for example using the `prometheusreceiver` receiver. I
- Processors: Can bundle/modify the metric/trace/log data as it comes through. The common `batchprocessor` groups and compresses data together to reduce load on exporters.
- Exporters: Exporters take the metric/trace/log data and export it to a downstream metric/trace/log solution. So for example, you might want to have trace data analysed with
AWS X-ray, but have log data go to Splunk. In the diagram above, we have metric data going to AWS CloudWatch, using the `awsemf` exporter.

# Arise the OTel Distro

> Now The OTel project release their own standard collector, which includes a bunch of sensible default receivers and exporters. However, what appears to be have happened is that a number of vendors have released expanded distributions of the collector, with additional bundled receivers and exporters to align with their target products. For example AWS has their [AWS Distro For Open Telemetry](https://github.com/aws-observability/aws-otel-collector) which includes receivers for various AWS container technologies, and exporters for X-ray and CloudWatch. SumoLogic [has a distribution](https://github.com/SumoLogic/sumologic-otel-collector) which supports exporting to SumoLogic (for both metrics and tracing). These specific distros are great for getting going quickly withina given ecosystem, but can cause issues if you want to route data to multiple providers.

Reviewer: ```suggestion
The OTel project released their own standard collector, which includes a bunch of sensible default receivers and exporters. However, what appears to be have happened is that a number of vendors have released expanded distributions of the collector, with additional bundled receivers and exporters to align with their target products. For example AWS has their [AWS Distro For Open Telemetry](https://github.com/aws-observability/aws-otel-collector) which includes receivers for various AWS container technologies, and exporters for X-ray and CloudWatch. SumoLogic [has a distribution](https://github.com/SumoLogic/sumologic-otel-collector) which supports exporting to SumoLogic (for both metrics and tracing). These specific distros are great for getting going quickly within a given ecosystem, but can cause issues if you want to route data to multiple providers.
```

> Other providers, such as [NewRelic](https://docs.newrelic.com/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/opentelemetry-setup) and [Honeycomb.io](https://docs.honeycomb.io/getting-data-in/opentelemetry-overview/#using-the-honeycomb-opentelemetry-endpoint) provide HTTP/GRPC endpoints that can directly receive data from the standard `otlpexporter` exporter, so you can just use the standard collector.

Reviewer: ```suggestion
Other providers, such as [NewRelic](https://docs.newrelic.com/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/opentelemetry-setup) and [Honeycomb.io](https://docs.honeycomb.io/getting-data-in/opentelemetry-overview/#using-the-honeycomb-opentelemetry-endpoint) provide HTTP/gRPC endpoints that can directly receive data from the standard `otlpexporter` exporter, so you can just use the standard collector.
```

# Agents, SDKs, Existing Libraries Oh-My

> Now that you understand how the collector works, the next step is getting your valuable application data into the collector. This is going to very much depend on the language your application is written in. In Java land, besides a manual SDK, there is support for an [agent-based automatic connector](https://opentelemetry.io/docs/instrumentation/java/automatic/), which out of the box has a lot of support for popular Java libraries. If you'd rather not use agent-based approach, the OpenTelemetry project has an alternative mechanism for injecting trace behaviour [specifically for SpringBoot](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/f1774cabe2153b7b118d3be664f0fa757b5e43d1/instrumentation/spring/spring-boot-autoconfigure/README.md).

Reviewer: ```suggestion
Now that you understand how the collector works, the next step is getting your valuable application data into the collector. This is going to very much depend on the language your application is written in. In Java land, besides a manual SDK, there is support for an [agent-based automatic connector](https://opentelemetry.io/docs/instrumentation/java/automatic/), which out of the box has a lot of support for popular Java libraries. If you would rather not use an agent-based approach, the OpenTelemetry project has an alternative mechanism for injecting trace behaviour [specifically for SpringBoot](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/f1774cabe2153b7b118d3be664f0fa757b5e43d1/instrumentation/spring/spring-boot-autoconfigure/README.md).
```

Reviewer: avoid conjunctions here and in other places.

        });
        webcontainer.addPortMappings({
            containerPort: 8080,
        });
        fargateTaskDefinition.addContainer("Otel", {
            image: aws_ecs.ContainerImage.fromRegistry('amazon/aws-otel-collector:latest'),
            logging: new aws_ecs.AwsLogDriver({ streamPrefix: 'ServiceBOtel', mode: aws_ecs.AwsLogDriverMode.NON_BLOCKING }),
            command: ['--config=/etc/ecs/ecs-default-config.yaml', '--set=service.telemetry.logs.level=DEBUG']
        });

        // // Create a load-balanced Fargate service and make it public
        let serviceB = new aws_ecs_patterns.ApplicationLoadBalancedFargateService(this, "SampleOtelService", {
            cluster: cluster,
            cpu: 512,
            desiredCount: 1,
            taskDefinition: fargateTaskDefinition,
            assignPublicIp: true, // this is purely to not require a NAT option.
            memoryLimitMiB: 1024,
            publicLoadBalancer: true,
            healthCheckGracePeriod: cdk.Duration.seconds(10),
        });
        dockerImageServiceB.repository.grantPull(serviceB.taskDefinition.obtainExecutionRole());

    }
}
```

> It's important to note, that X-ray currently requires a specific time based ID format, which Quarkus discuss in [this section](https://quarkus.io/guides/opentelemetry#id-generator). You effectively need to use the X-Ray IDGenerator, or nothing will appear in the X-ray console.

Reviewer: ```suggestion
It's important to note, that X-ray currently requires a specific time based ID format, which Quarkus discusses in [this section](https://quarkus.io/guides/opentelemetry#id-generator). You effectively need to use the X-Ray IDGenerator, or nothing will appear in the X-ray console.
```

The X-ray console looks like the following. This example demonstrates hitting a sample `/hello` endpoint, via the deployed load-balancer, hitting the Quarkus application, and then making a `HTTP GET` call to a Star Wars test endpoint: `https://www.swapi.tech/api/starships/3`.

![x-ray-console](https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2022/11/opentelemetry_xray.png)

# Why/When Would I Use It?

> It's a great way of integrating metric and trace monitoring into new applications, or applications that don't have existing solutions in this space. I've excluded logging for the moment as the specification is still in a [draft status](https://opentelemetry.io/docs/reference/specification/status/) at the time of writing, but it is certainly an exciting development. This is a pretty fast moving space, which will mature and solidify over time, but it will be the way of integrating vendor-agnostic tracing/metric/logging functionality into your application moving forward.

Reviewer: Add a conclusion and a call to action after this final section.

Author: I've added a conclusion just below this. Hopefully that makes it a little cleaner.
